lib/spack/spack/environment/environment.py

```diff
@@ -1794,31 +1797,73 @@ def _add_concrete_spec(self, spec, concrete, new=True):
         self.concretized_order.append(h)
         self.specs_by_hash[h] = concrete
 
-    def _get_overwrite_specs(self):
-        # Find all dev specs that were modified.
-        changed_dev_specs = [
-            s
-            for s in traverse.traverse_nodes(
-                self.concrete_roots(), order="breadth", key=traverse.by_dag_hash
-            )
-            if _is_dev_spec_and_has_changed(s)
-        ]
+    def _get_overwrite_specs(self, _database=None):
+        """Gather all specs that should be overwritten. There are three instances
+        where this is the case:
+
+        * The spec is a "develop" spec, and a user has (outside of spack)
+          modified one of the source files.
+        * Any parent of such a spec.
+        * A dependency that is a "develop" spec was installed more-recently
+          this can occur if, for example, x->y, you change and reinstall
+          the in-development spec y, and then later reinstall the whole
+          environment.
+        """
+        db = _database or spack.store.STORE.db
 
-        # Collect their hashes, and the hashes of their installed parents.
-        # Notice: with order=breadth all changed dev specs are at depth 0,
-        # even if they occur as parents of one another.
-        return [
-            spec.dag_hash()
-            for depth, spec in traverse.traverse_nodes(
-                changed_dev_specs,
-                root=True,
-                order="breadth",
-                depth=True,
-                direction="parents",
-                key=traverse.by_dag_hash,
-            )
-            if depth == 0 or spec.installed
-        ]
+        overwrite_specs = set()
+        transitive_dev_install_times = collections.defaultdict(int)
+
+        def update_transitive_dev_install_times(spec):
+            dev_path_var = spec.variants.get("dev_path", None)
+            if dev_path_var and db.installed(spec):
+                _, record = db.query_by_spec_hash(spec.dag_hash())
+                when_this_spec_was_installed = record.installation_time
+
+                transitive_dev_install_times[spec] = max(
```

Comment:
Why don't you assign `transitive_dev_install_time_child = transitive_dev_install_times[spec]` first and then update it, instead of calling `transitive_dev_install_times[spec]` 3 times?